internal/app/secretless/handlers/pg/protocol/ssl.go

```diff
@@ -0,0 +1,149 @@
+
+	switch mode := o["sslmode"]; mode {
+	case "disable":
+		tlsConf.UseTLS = false
+		return tlsConf, nil
+		// "require" is the default.
+	case "", "require":
+		// We must skip TLS's own verification since it requires full
+		// verification since Go 1.3.
+		tlsConf.InsecureSkipVerify = true
+
+		// From http://www.postgresql.org/docs/current/static/libpq-ssl.html:
+		//
+		// Note: For backwards compatibility with earlier versions of
+		// PostgreSQL, if a root CA file exists, the behavior of
+		// sslmode=require will be the same as that of verify-ca, meaning the
+		// server certificate is validated against the CA. Relying on this
+		// behavior is discouraged, and applications that need certificate
+		// validation should always use verify-ca or verify-full.
+		if len(o["sslrootcert"]) > 0 {
+			tlsConf.VerifyCaOnly = true
+		}
+	case "verify-ca":
+		// We must skip TLS's own verification since it requires full
+		// verification since Go 1.3.
+		tlsConf.InsecureSkipVerify = true
+		tlsConf.VerifyCaOnly = true
+	//case "verify-full":
+	//	tlsConf.ServerName = o["host"]
+	default:
+		// TODO add verify-full below
+		return TLSConfigWrapper{}, fmt.Errorf(`unsupported sslmode %q; only "require" (default), "verify-ca", and "disable" supported`, mode)
+	}
+
+	return tlsConf, nil
+}
+
+// HandleSSLUpgrade upgrades a net.Conn using TLSConfigWrapper
+func HandleSSLUpgrade(connection net.Conn, tlsConf TLSConfigWrapper) (net.Conn, error) {
+	err := sslClientCertificates(&tlsConf.Config, tlsConf.Options)
+	if err != nil {
+		return nil, err
+	}
+	err = sslCertificateAuthority(&tlsConf.Config, tlsConf.Options)
+	if err != nil {
+		return nil, err
+	}
+
+	// Accept renegotiation requests initiated by the backend.
+	//
+	// Renegotiation was deprecated then removed from PostgreSQL 9.5, but
+	// the default configuration of older versions has it enabled. Redshift
+	// also initiates renegotiations and cannot be reconfigured.
+	// TODO: what does mysql require for this ?
+	tlsConf.Renegotiation = tls.RenegotiateFreelyAsClient
+
+	client := tls.Client(connection, &tlsConf.Config)
+	if tlsConf.VerifyCaOnly {
+		err := sslVerifyCertificateAuthority(client, &tlsConf.Config)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+// sslClientCertificates adds the certificate specified in the "sslcert" and
+// "sslkey" settings
+func sslClientCertificates(tlsConf *tls.Config, o options) error {
+	// The client certificate is only loaded if the setting is not blank.
+	sslcert := o["sslcert"]
+	if len(sslcert) == 0 {
+		return nil
+	}
+
+	sslkey := o["sslkey"]
+
+	certPEMBlock := []byte(sslcert)
+	keyPEMBlock := []byte(sslkey)
+
+	cert, err := tls.X509KeyPair(certPEMBlock, keyPEMBlock)
+	if err != nil {
+		return err
+	}
+
+	tlsConf.Certificates = []tls.Certificate{cert}
+	return nil
+}
+
+// sslCertificateAuthority adds the RootCA specified in the "sslrootcert" setting.
+func sslCertificateAuthority(tlsConf *tls.Config, o options) error {
```

Comment:
Since this is not returning a CertificateAuthority, and this implicitly does validation, I think `addRootCA`would be a more accurate name.